dts/bindings/usb-c/usb-c-connector.yaml

```diff
@@ -0,0 +1,173 @@
+# Copyright 2022 The Chromium OS Authors
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    A USB Type-C connector node represents a physical USB Type-C connector.
+    It should be a child of a USB-C interface controller or a separate node
+    when it is attached to both MUX and USB-C interface controller.
+
+    This is based on Linux, documentation:
+      https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/devicetree/bindings/connector/usb-connector.yaml?h=v5.19&id=3d7cb6b04c3f3115719235cc6866b10326de34cd
+
```

Comment:
Looks like we are missing the `ports` structure from the linux binding.  Any reason for this?

Reply:
The ports defined in Linux have a slightly different representation. For example, it defines ports as follows:
```
     port@0:
        description: High Speed (HS), present in all connectors.

      port@1:
        description: Super Speed (SS), present in SS capable connectors.

      port@2:
        description: Sideband Use (SBU), present in USB-C. This describes the
          alternate mode connection of which SBU is a part.
```
I thought this was to restrictive and decided to leave the port definition up to the user.

Reply:
Does one of these not map to what you are trying to describe?

(as above, we should be trying to match the linux binding as much as possible.  If one of these port@N doesn't match what you need, than I think its reasonable to add a new port@3, port@4, etc..).  But we should match the structure if we can.

Reply:
The Linux port combine USB and USB-C. Currently, in zephyr, the two busses are separate. port@0 refers to USB 2.0 and port@1 refers to USB 3.0. The Sideband Use (SBU) on port@2 is also not used in the stack.

Combining USB and USB-C is probably something we should discuss in an Issue and implement in a separate PR.

Reply:
What is the actual HW configuration we are trying to describe?

Reply:
We can only implement a USB-C Sink Power Delivery device and each port should have a Type-C Port Controller (TCPC) and a way to measure USB-C VBUS (Some TCPCs can be used to measure VBUS).

```
port1: usbc-port@1 {
                        compatible = "usb-c-connector";
                        reg = <1>;
                        tcpc = <&ucpd1>;
                        vbus = <&vbus1>;
                        power-role = "sink";
                        sink-pdos = <PDO_FIXED(5000, 100, 0)>;
                };
```

Reply:
I don't know enough about usb to parse that :).

Put the actual software aside, is it possible to describe the HW we want to support using the linux binding or not?  If so, what would that dts (nodes) look like (based purely on the linux binding)?

Reply:
The short answer is No. The Linux ports are mostly concerned with USB data and not USB-C power delivery.

Reply:
What's the disconnect between the two?  Seems like the binding is trying to describe both cases for a usb-connector.   Or are you saying that specifically with regards to `port@N`?

Reply:
I'm glancing at some DTS files in linux and see things like:

```
        ptn5110: tcpc@50 {
                compatible = "nxp,ptn5110";
                pinctrl-names = "default";
                pinctrl-0 = <&pinctrl_typec1>;
                reg = <0x50>;
                interrupt-parent = <&gpio2>;
                interrupts = <11 IRQ_TYPE_LEVEL_LOW>;
                status = "okay";

                port {
                        typec1_dr_sw: endpoint {
                                remote-endpoint = <&usb1_drd_sw>;
                        };
                };

                typec1_con: connector {
                        compatible = "usb-c-connector";
                        label = "USB-C";
                        power-role = "dual";
                        data-role = "dual";
                        try-power-role = "sink";
                        source-pdos = <PDO_FIXED(5000, 3000, PDO_FIXED_USB_COMM)>;
                        sink-pdos = <PDO_FIXED(5000, 3000, PDO_FIXED_USB_COMM)
                                     PDO_VAR(5000, 20000, 3000)>;
                        op-sink-microwatt = <15000000>;
                        self-powered;
                };
        };
};
```

or
```
        wusb3801: tcpc@60 {
                compatible = "willsemi,wusb3801";
                reg = <0x60>;
                interrupt-parent = <&gpio0>;
                interrupts = <RK_PA4 IRQ_TYPE_LEVEL_LOW>;
                pinctrl-0 = <&tcpc_int_l>;
                pinctrl-names = "default";

                connector {
                        compatible = "usb-c-connector";
                        label = "USB-C";
                        vbus-supply = <&otg_switch>;
                        power-role = "dual";
                        try-power-role = "sink";
                        data-role = "dual";
                        typec-power-opmode = "default";
                        pd-disable;

                        ports {
                                #address-cells = <0x1>;
                                #size-cells = <0x0>;

                                port@0 {
                                        reg = <0x0>;

                                        typec_hs_usb2phy0: endpoint {
                                                remote-endpoint = <&usb2phy0_typec_hs>;
                                        };
                                };
                        };
                };
        };
```

Reply:
From the Linux ports description:
```
description: OF graph bindings modeling any data bus to the connector
      unless the bus is between parent node and the connector. Since a single
      connector can have multiple data buses every bus has an assigned OF graph
      port number as described below.
```
From the USB-C stacks port of view, it doesn't care about USB data bus or  Sideband SBU connections.

The USB-C stack in this PR only cares about Sink Power Delivery, not anything USB

Reply:
From the example you provided
```
port@0 {
                                        reg = <0x0>;

                                        typec_hs_usb2phy0: endpoint {
                                                remote-endpoint = <&usb2phy0_typec_hs>;
                                        };
                                };
```
This is USB not USB-C.

Reply:
I'll add the Linux ports property but it won't be used by the USB-C stack. Could I do the addition in a separate PR?

Reply:
the devicetree is trying to describe hardware.  So lets ignore what the various software stacks do or don't do with that data.

In the examples I provided are `tcpc@50` and `tcpc@60` the type `tcpc` controller?

What I'm trying to focus on are the `tcpc` and `vbus` properties.  I think I understand how `port` isn't relevant at this time for you for what you need described.

Reply:
@keith-zephyr if the diagram, is the gray pill box the 'usb-c-connector'?

Reply:
Yes, that is correct.

Reply:
The Linux USB Connector node is largely defining USB policy information.

Does it make sense for Zephyr to create a "usb-c-pd" node type to specify the chips (TCPC, PPC, MUXes), and to make the USB Connector a child?

```
	ports {
		#address-cells = <1>;
		#size-cells = <0>;
		/* usbc.rst usbc-port start */
		port1: usbc-port@1 {
			compatible = "usb-c-pd";
			reg = <1>;
			tcpc = <&ucpd1>;
			/* Note - ppc/muxes not part of this PR but needed later */
			ppc = <&ppc1>;
			muxes = <&retimer1 &mux1>;
			connector {
			    compatible = "usb-c-connector";
			    power-role = "dual";
			    vbus-supply = <&vbus1>;
			    ... /* Other properties: vdos, data-role, etc */ 
			};
		};
	};
```

Reply:
Here are some links to the commits on the linux side that seem associated with google:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/commit/Bindings/connector/usb-connector.yaml?id=d76a3c08d6496674da8dbb17e95546852f60e0ea
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/commit/Bindings/connector/usb-connector.yaml?id=f28dc9ecd6697420b0777a99a5d1bc97f1a0d857
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/commit/Bindings/connector/usb-connector.yaml?id=9af0f74778519a0f8bf22eddf5ad6c87b1223321